service/src/main/java/com/codeforcommunity/processor/ReservationProcessorImpl.java

```diff
@@ -0,0 +1,300 @@
+package com.codeforcommunity.processor;
+
+import static org.jooq.generated.Tables.*;
+
+import com.codeforcommunity.api.IReservationProcessor;
+import com.codeforcommunity.auth.JWTData;
+import com.codeforcommunity.dto.reservation.*;
+import com.codeforcommunity.enums.PrivilegeLevel;
+import com.codeforcommunity.enums.ReservationAction;
+import com.codeforcommunity.enums.TeamRole;
+import com.codeforcommunity.exceptions.*;
+import java.sql.Timestamp;
+import java.util.Optional;
+import org.jooq.DSLContext;
+import org.jooq.generated.tables.records.*;
+
+public class ReservationProcessorImpl implements IReservationProcessor {
+
+  private final DSLContext db;
+
+  public ReservationProcessorImpl(DSLContext db) {
+    this.db = db;
+  }
+
+  public Optional<ReservationsRecord> lastAction(int blockId) {
+    return Optional.ofNullable(
+        db.selectFrom(RESERVATIONS)
+            .where(RESERVATIONS.BLOCK_ID.eq(blockId))
+            .orderBy(RESERVATIONS.PERFORMED_AT.desc())
+            .limit(1)
+            .fetchOne());
+  }
+
+  private void basicChecks(int blockId, Integer userId, Integer teamId) {
+    if (!db.fetchExists(db.selectFrom(BLOCKS).where(BLOCKS.ID.eq(blockId)))) {
+      throw new ResourceDoesNotExistException(blockId, "block");
+    }
+
+    if (userId != null) {
```

Comment:
This is nitpicky but all of the double nested if clauses you have in this file can be combined into a single if. Java has boolean shortcircuiting so if userId is equal to null then the second clause won't execute so there's no chance of a Null Pointer Exception

Reply:
Did most of these